### PR TITLE
Platformio project

### PR DESCRIPTION
**Pull request by janLo**
This converted the project to a platformio project for easier distribution/dependency-management/build. It can easily buit with
```
pio run
```
This takes care of all library dependencies and downloads missing ones automatically. There are multiple board definitions that are available as target envs. You can select a specific env using 
```
pio run --environment <env>
```
where `env` can be `esp12e`, `d1_mini` or `nodemcuv2`.

Platformio is also able to generate "project"-descriptions for several other development tools like clion, vscode, vim, visualstudio, etc. Just run `pio init --ide <myide>`.

### CODE REVIEW

File: src/MQTT433gateway.ino [hunk 1]
```diff
@@ -66,7 +66,7 @@ boolean rawMode = false;
 String otaURL = "";
 
 void setup() {
-  Serial.begin(115200,SERIAL_8N1,SERIAL_TX_ONLY); 
+  Serial.begin(115200,SERIAL_8N1,SERIAL_TX_ONLY);
```
Comment (puuu) [inline src/MQTT433gateway.ino]: Please also fix the missing spaces after the comma. (maybe better in a separate commit)

File: lib/Heartbeat/HeartbeatFlashing.cpp [hunk 2]
```diff
@@ -27,6 +27,8 @@
   SOFTWARE.
 */
 
+#ifdef HB_FLASHING
+
```
Comment (puuu) [inline lib/Heartbeat/HeartbeatFlashing.cpp]: Why we need this, where is it used?

File: README.md [hunk 3]
```diff
@@ -133,7 +148,7 @@ binary_sensor:
 ```
 
 The message of a TCM 218943 weather station sensor looks like this:
-`rf434/recv/tcm/108 {"id":108,"temperature":22.7,"humidity":50,"battery":1,"button":0}`. 
+`rf434/recv/tcm/108 {"id":108,"temperature":22.7,"humidity":50,"battery":1,"button":0}`.
```
Comment (puuu) [inline README.md]: Thanks

File: .gitignore [hunk 4]
```diff
@@ -1,3 +1,4 @@
-*~
-*.bin
-net-passwd.h
```
[the remark this answers is not in the document]
Comment (janLo) [reply .gitignore]: They're not needed for a platformio project, as the build happens in a separate build directory.

File: lib/Heartbeat/HeartbeatFlashing.cpp (hunk 2 above)
Comment (janLo) [reply lib/Heartbeat/HeartbeatFlashing.cpp]: Platformio tries to build the object file but got somehow confused as this code is not used abywhere. So this is for now basically a `#if 0`.

**Comment by janLo** [comment]
Thank you for your feedback. I'm aware that this somehow "breaks" the arduino compatibility. Sadly they're not really compatible with all the other build/development tools. But I think this is worth it, as the PlatformIO environment is generally easier to use. You don't need to install the IDE and fiddle with the UI to get the board right and all the libs available. You just install and run it.

I'm also planning to implement more features like "interactive" configuration via some sort of UI to make it possible to just distribute prebuild binaries that the user can then just configure on the device without the need to build anything. For this a proper library-dependency-management is a mayor benefit. You just specify the dependencies and Platformio takes care of getting, configuring and building it.

Also board-specific options like Pin-selection, etc is easier to describe with platformio as you have distinct environments for this. Supporting a "new" board (or adding one like `huzzah`) is just a new block in the `platformio.ini`.

Also, with PlatformIO you get a IDE-Agnostic build-system. You can generate configurations for most available IDEs (unfortunate except the Arduino IDE).

I'll try to clean up the changes as you suggested on a best-effort basis.

**Comment by janLo** [comment]
I cleaned the commits up as requested. The discussion about PlatformIO or not is still not done though.

**Comment by puuu** [comment]
How about, having a directory for the PlatformIO project and link the files by symbolic links?

**Comment by torwag** [comment]
Symbolic links do not work afaik under git.

**Comment by torwag** [comment]
My opinion to this: If the readme guides a newbie through the entire build process, it doesn't matter if it is based on Arduino IDE or PlatformIO, at least from a newbie point of view.
Thus, if it is a exclusive or decision, the only point which is important is maintenance and benefits on the developer side. If PlatformIO provides benefits here, we have to take care to provide a very good documentation for newbies and can profit from a better build environment. If there is no benefit on the developer side, there is no need to change it.

**Comment by puuu** [comment]
My tests with git on linux support links. By cloning a repository, the links will be created. How it works on Windows, I am not sure.

**Comment by janLo** [comment]
I think its not worth the effort to support both, as it means, that you need to test both all the time, maintain documentation for both, etc. Especially if the project evolves and get more external dependencies this might be a growing hassle.

**Comment by puuu** [comment]
> we have to take care to provide a very good documentation for newbies 

@torwag, do you think, the introduced changes in bd885bc2570b50bbcd2f88c6cac4797a2848f97a are not sufficient for a newbies?

Beside the pure newbies, there are also the want-to-be-a-make-and-attended-a-workshop-about-arduino guys that already have a running Arduino IDE.

> as it means, that you need to test both all the time

That is why we use travis CI, we only need to integrate the Arduino IDE.

On the other side, if we went the whole road with configuration frontend, as started in #18. We only need to provide binary releases for newbies and can profit from a better build environment. But this is something I can not maintain by myself.

**Comment by torwag** [comment]
Hi,
there was a small typo which for which I just send in a patch. Other then this, I think the readme works out well. We might need to add what kind of boards people can choose from more clearly (huzzah is still missing) and a bit more details how to flash the board. E.g. how to address a specific serial port could be added. As platformio is completely command line based, we could make it even easier and provide a little script to 
1. generate the net-passwd.h file
2. select the board
3. select the serial port
4. compile and flash
The trick here would be to find something which works OS agnostic.
Having a script like this, all and every newbie should be happy. 
Then it comes down to how easy it is to install platformIO on a particular system. As Linux guy, I am super spoiled by superb package managers... I have no idea how platformio installation works out on Windows or Mac.

Having a webfront-end and a suitable set-up procedure, in the future, indeed there is no need to take care much about newbies for the tool-chain side of view, as the project could provide binaries and perform OTA updates.

If the Arduino support should not be dropped, I would prefer to have it the other way around. E.g. create a folder including an Arduino environment based on the underlying platformio set-up. 

As for why switching to PlatformIO? In my opinion PlatformIO is actually even easier to use for newbies, as one has not to add repositories, and install packages within a GUI (which might come even in a different languages as described in the readme). The entire build-process could be condensed down in a single command line. In addition, we get free dependency checks, thus, the chance that people shoot themself in the foot by using outdated libs or strange combinations, is smaller and we get a toolchain, which makes it easy to port the project to other boards and platforms as well.   

Thus, I say, lets pimp the description a bit more and move over to PlatformIO.

**Comment by torwag** [comment]
I started a fresh build and noticed all jotted down where a newbie might face challenges. I then modified the README to add those parts and submitted a PR to @janLo,  which imho improves the README for a newbie sufficiently. With this, I guess we could switch to platformIO. Only thing I did not test... how easy it goes under Windows or Mac, as I lack both OSes (my wife will not let me use her PC to infest it with some of my tinkerer stuff ;) )
BTW. The other projects we were looking into for web-management use both platformIO and other prominent projects in this area do the same. I guess they all do for a reason ;)

**Comment by puuu** [comment]
Thank you for all the comments and work. Since there are no new comments, my decision is like follow:

Since this project is for advanced users, e.g., they need electronic skills to build the gateway, they need a MQTT brocker and they need a configured home automation system. We can consider that the user is able to install platfomIO. Furthermore, we should provide binary releases and a documentation how to flash them.

If we provide full configuration via web interface, binary releases are fine for users, other projects do the same. Developer can benefit from a powerful development environment.

As for the Arduino IDE compatibility, I will backport PR #14, #15 and maybe #17. Then I create an `arduino` branch, for further minor changes (e.g. #20). 
The readme of the `arduino` branch will get a depredation warning.  The `master` branch will get this pull request and the readme should mention the `arduino` branch.

**Comment by torwag** [comment]
Sound good good to me. Let's dot it, as this will help to get all the other stuff going on more smoothly

**Comment by puuu** [comment]
For the arduino branch this is done in 85121fc74af4cd3d02d5f4e6ef6de594d363e719 .

**Comment by puuu** [comment]
We did it!
@janLo and @torwag thank you for the discussion and the commits.

Merged in acbe8cc9d0e4f371294ff7998897e5bfcbafcf48...75558c1dec9d956c1c061131baf9fdf0c67ad5c5 .

**Pull request closed without merging**